ldk-server/src/service.rs

```diff
@@ -29,51 +25,66 @@ use crate::api::open_channel::{handle_open_channel, OPEN_CHANNEL_PATH};
 use crate::api::update_channel_config::{
 	handle_update_channel_config_request, UPDATE_CHANNEL_CONFIG_PATH,
 };
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
 
 #[derive(Clone)]
 pub struct NodeService {
-	node: Arc<Node>,
+	context: Arc<Context>,
 }
 
 impl NodeService {
 	pub(crate) fn new(node: Arc<Node>) -> Self {
-		Self { node }
+		Self { context: Arc::new(Context { node }) }
 	}
 }
 
+pub(crate) struct Context {
+	pub(crate) node: Arc<Node>,
+}
```

Comment:
Could you remind me why `PaginatedKvStore` can't be in `NodeService`? Is it just to save passing two parameters to the relevant handlers? If this is only need for one handler, then I'm not sure if we should bother adding the wrapper.

Reply:
Hmm... seems less complicated if we don't add another struct. I don't see it as a circular dependency. The only reason why these functions aren't methods on `NodeService` is to get around the fact that `Service` uses `Future` instead of being an `async` trait. Using `Arc<NodeService>` is just like a `&self` parameter that works with `async`.

The wrapper is only needed because both `Service` and `Arc` don't belong to this crate, so we need another type to implement `Service` on. But that is only needed locally.

Reply:
Here's the first part of the change. Then you just need to update the handlers to take `Arc<NodeService>` instead of `Arc<Node>`.

```diff
diff --git a/ldk-server/src/main.rs b/ldk-server/src/main.rs
index 6d31341..010a40f 100644
--- a/ldk-server/src/main.rs
+++ b/ldk-server/src/main.rs
@@ -3,7 +3,7 @@ mod io;
 mod service;
 mod util;
 
-use crate::service::NodeService;
+use crate::service::{ArcNodeService, NodeService};
 
 use ldk_node::{Builder, Event, LogLevel};
 
@@ -76,6 +76,8 @@ fn main() {
                node.config().listening_addresses.as_ref().unwrap().first().unwrap()
        );
 
+       let node_service = Arc::new(NodeService::new(Arc::clone(&node)));
+
        runtime.block_on(async {
                let mut sigterm_stream = match tokio::signal::unix::signal(SignalKind::terminate()) {
                        Ok(stream) => stream,
@@ -118,7 +120,7 @@ fn main() {
                                        match res {
                                                Ok((stream, _)) => {
                                                        let io_stream = TokioIo::new(stream);
-                                                       let node_service = NodeService::new(Arc::clone(&node));
+                                                       let node_service = ArcNodeService(Arc::clone(&node_service));
                                                        runtime.spawn(async move {
                                                                if let Err(err) = http1::Builder::new().serve_connection(io_stream, node_service).await {
                                                                        eprintln!("Failed to serve connection: {}", err);
diff --git a/ldk-server/src/service.rs b/ldk-server/src/service.rs
index 1062cbc..fa1c526 100644
--- a/ldk-server/src/service.rs
+++ b/ldk-server/src/service.rs
@@ -8,6 +8,7 @@ use hyper::{Request, Response, StatusCode};
 use prost::Message;
 
 use std::future::Future;
+use std::ops::Deref;
 use std::pin::Pin;
 use std::sync::Arc;
 
@@ -41,13 +42,15 @@ impl NodeService {
        }
 }
 
-impl Service<Request<Incoming>> for NodeService {
+pub(crate) struct ArcNodeService(pub Arc<NodeService>);
+
+impl Service<Request<Incoming>> for ArcNodeService {
        type Response = Response<Full<Bytes>>;
        type Error = hyper::Error;
        type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
 
        fn call(&self, req: Request<Incoming>) -> Self::Future {
-               let node = Arc::clone(&self.node);
+               let node = Arc::clone(&self.0.deref().node);
                // Exclude '/' from path pattern matching.
                match &req.uri().path()[1..] {
                        GET_NODE_INFO => Box::pin(handle_request(node, req, handle_get_node_info_request)),
@@ -94,7 +97,7 @@ async fn handle_request<
        F: Fn(Arc<Node>, T) -> Result<R, ldk_node::NodeError>,
 >(
        node: Arc<Node>, request: Request<Incoming>, handler: F,
-) -> Result<<NodeService as Service<Request<Incoming>>>::Response, hyper::Error> {
+) -> Result<<ArcNodeService as Service<Request<Incoming>>>::Response, hyper::Error> {
        // TODO: we should bound the amount of data we read to avoid allocating too much memory.
        let bytes = request.into_body().collect().await?.to_bytes();
        match T::decode(bytes) {
```

Reply:
> Moreover node_service could have members which are not directly required for request-handling such as `authorizer`, and we wouldn't want to expose them as it is to handlers. Imo it would be better to have separation of concerns here.

We can always change it when that is the case. Handlers are simply methods on `NodeService`, IMO.

Reply:
Not sure I really followed the example that you gave. Currently, `NodeService` is _only_ used for handling. Do you expect it to be used for something else (i.e., not handling) that would require an authorizer?

Reply:
Was thinking about this more. Assuming an authenticator would be needed *before* handling, then it seems we can create `Context` inside `handle_request` rather than storing it in `NodeService`. Then it wouldn't even need to be an `Arc` since it could be passed by value to the handler. I'd be fine with that. As the code is currently written, a `NodeService` is composed of a `Context`, which doesn't seem right.

Reply:
Right, inside `call` is accurate. I misspoke earlier.

Reply:
Err... actually you need to create `Context` in `handle_request` to pass it to `call`, IIUC, since you need both the `Arc<Node>` and `Arc< PaginatedKVStore>` from `NodeService` to create `Context`.

Reply:
Yeah, that looks good. Seems I was just mixing up the names for some reason. 😛